docs/cardano-monads.rst

```diff
@@ -0,0 +1,922 @@
+* Extensibility: MODERATE. Defining new effects boils down to declaring a method
+  record and a ``Has`` class for it. Then, for each field in the method record we
+  must define a function that extracts this method using the ``Has`` class and
+  runs it. Boilerplate-heavy, yes, but at least effects don't need to know about
+  each other.
+
+* Ease of use: MODERATE. Beginners can quickly grasp the basic concepts, but
+  there are pitfalls when it comes to lifting method records (see section on
+  ``SendActions``).
+
+* Compile-time performance: SUPERB. There's nothing to it, GHC can easily handle
+  this style of code.
+
+* Run-time performance: MODERATE. A flat ``ReaderT`` layer is good, but method
+  records are stored within the context (rather than passed via instance
+  search). Because of this, the compiler can't assume that the method record is
+  coherent (it's not, you can modify it with ``local``), and it limits inlining.
+  Instance specialization can't be performed because there are no instances to
+  speak of.
+
+* Run-time configurability: SUPERB. You can put just about anything into these
+  manually defined method records, modify them at will, etc. However, we need to
+  figure out a good story for method records that depend on other method
+  records, because it's the same pitfall as with ``SendActions``: how can we
+  ensure that if we change something (say, a logging method) with ``local``,
+  other method records that use logging will be updated accordingly?
+
+* Predictability: LOW/MODERATE (depends on whether we have a ``newtype`` over
+  ``ReaderT``). In case there's no newtype, ``runReaderT`` can change the
+  behavior, which is counter-intuitive. In case there's a newtype, there's too
+  much freedom with regards to what can be put into method records.
+
+Verdict: the approach is viable but has its costs.
+
+
+Dictionary-Passing Style
+~~~~~~~~~~~~~~~~~~~~~~~~
+
+Similar to previous approach and one was widely used by the networking team.
+
+As we understand ideas behind it:
+
+- Single base monad
+
+  - Similarly to ``ReaderT IO`` approach
+
+- Differentiating implementations all stored in records
+
+  - No type classes and instances like ``instance MyClass (ReaderT Ctx IO)``
+  - Each function accepts a single **explicit** parameter of type ``Ctx m``
+  - This type ``Ctx`` contains all needed methods
+
+Point-by-point rundown:
+
+* Flexibility: MODERATE. (Same as above).
+
+* Extensibility: GOOD. (Same as above, minus a bit of boilerplate helpers).
+
+* Ease of use: LOW. (Same as above, plus one needs to manually pass the
+  context everywhere).
+
+* Compile-time performance: SUPERB. (Same as above).
+
+* Run-time performance: MODERATE. (Same as above).
+
+* Run-time configurability: SUPERB. (Same as above)
+
+* Predictability: MODERATE. Checking what implementation is used is easy as tracing
+  the provenance of the input context, but there's too much freedom with regards to
+  what can be put into the method records.
```

Comment:
I don't agree with this. Records have types just like class methods do. There's no difference in freedom.

Reply:
I articulated this point poorly. The idea is that when the behavior is defined by instances for a base monad in a single module, it's easy to see in what combination the implementations are used (assuming we don't use transformers to modify the behavior!) With records anything can be passed anywhere.

Reply:
This decrease in predictability is the flip side of increase in extensibility/flexibility.

Reply:
Ok. I should mention then that the dynamic binding solution we've been discussion must score very low for predictability :)

Reply:
Yes, sure, it does score lower for the same reasons.

You might also argue that it scores even lower because changing one capability might affect other capabilities in a ways the developer might not have anticipated — but I have a counter-argument here. With regular method records the developer might not know that some effect is trapped in another effect's closure, this would lead to unpredictable *lack* of change, and tracing closure is harder still. Thus I feel that late binding or closure-using method records score approximately the same on the predictability scale.

Reply:
> You might also argue that it scores even lower because changing one capability might affect other capabilities in ways the developer might not have anticipated

Yes precisely this.

> I have a counter-argument here. With regular method records the developer might not know that some effect is trapped in another effect's closure, this would lead to unpredictable lack of change

Ah but this is what we all expect isn't it? If I abstract over a `DB m`, I'm telling you that I can do what I need with *any* `DB m`. I don't care what's in its closure, I don't care what logging implementation it uses, and if I did, *I'd say so myself*, by abstracting over a `(Logging m -> DB m)` instead.

"unpredictable lack of change" is nonsense in my opinion. Pure functional programming is all about lack of change, and it's very predictable: nothing happens, with probability 1.

Reply:
> "unpredictable lack of change" is nonsense in my opinion

If a function is meant to override some behavior and it only overrides a part of it, I consider it an unpredictable lack of change. If you had `map :: (a -> a) -> f a -> f a` that only modifies half of the elements, it's an unpredictable lack of change.

I would like to remind that the late binding strategy is purely functional as well, and there are no changes with probablity 1. I don't see how this appeal to immutability is valid.

Reply:
> Ah but this is what we all expect isn't it?

No :)

Reply:
> If a function is meant to override some behavior and it only overrides a part of it, I consider it an unpredictable lack of change. If you had map :: (a -> a) -> f a -> f a that only modifies half of the elements, it's an unpredictable lack of change.

It's not a good analogy. `map f xs` doesn't change `xs`.

> I would like to remind that the late binding strategy is purely functional as well, and there are no unrequested changes with probablity 1. I don't see how this appeal to immutability is valid.

Yes we speak of dynamic/late binding but of course the implementation of it is in the end pure functional and nothing is truly dynamically bound. There's no true mutability, I'm just saying that it can sure feel like mutability.

Also, even in this solution it's possible to avoid the "dynamic binding" by putting an antecedent implementation into the closure of another implementation. To get the "dynamic binding" feel you have to ensure that the implementations you're using subscribe to the "dynamic binding" by pulling their antecedent implementations from the reader context rather than from elsewhere. In some sense, it strictly increases unpredictability: you get the unpredictable change (I didn't know the DB's logging would change?!?!) and also the so called "unpredictable lack of change" (which I don't actually endorse as a real thing) when the DB sidesteps the dynamic binding trick and doesn't use your new logging implementation.

Reply:
> It's not a good analogy. map f xs doesn't change xs.

Well, neither does `withReaderT`.

> when the DB sidesteps the dynamic binding trick and doesn't use your new logging implementation.

A faulty implementation.

> you get the unpredictable change (I didn't know the DB's logging would change?!?!) 

Well, who wouldn't expect logging to change by a function call like this?

```
runNoLoggingT $ do
  ...
```

or like this?

```
withReaderT (overrideCap loggingDummy) $ do
  ...
```

Just read the code. In the first case it says: RUN WITH NO LOGGING. That's the name of the transformer runner from `monad-logger`.

In the second case it says: OVERRIDE LOGGING TO DUMMY. That's just reading it like English.

Of course, we can consider a developer well aware and thinking thoroughly about what's actually happening to the transformer stack in case of `runNoLoggingT`, or to explicit dictionaries with closures — you are such a developer. For you it's 100% predictable. (But maybe not all implications are clear, otherwise why would you propose the capabilities idea?)

But for me it wasn't predictable before a few days ago, when I realized this situation with closures. I simply assumed that `runNoLoggingT` removes logging, because I didn't think about it thoroughly and because it's what the name suggests.

I could've in the same manner expected that passing `noLogging` dictionary would disable logging in the function I'm calling, unless I know the ins and outs of method dictionary passing, about its caveats.

To summarize: late binding and closures might be equally unpredictable depending on developer's expectation. Late binding is *more* predictable in the sense that it changes the behavior uniformly (oh no, I sense the "doesn't change anything" nitpick — okay, the new passed capability is used coherent in the subcomputation).

Reply:
> > when the DB sidesteps the dynamic binding trick and doesn't use your new logging implementation.
> A faulty implementation.

Yet also completely legitimate. Suppose I want all queries from my `DB m` to be logged to my special log server. I build it into the closure, using a `Logging m`. I don't want you to be able to change this, and indeed you can't, and that's good. Will you complain that this is unexpected, when you swap in a new logger but my `DB m` still logs to my server?